### PR TITLE
feat: feat/implement_0x90_SWAP1_Opcode

### PR DESCRIPTION
**Pull request by Noeljarillo**
Implementation of Swap_i and swap_n Opcodes

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### CODE REVIEW

**Comment by Noeljarillo** [comment]
Now hope is all good! still need to add more edge case tests like old cairo implementation, but wanted to check all is good before, Ive followed PUSH implementation

File: crates/evm/src/instructions/exchange_operations.cairo [hunk 1]
```diff
@@ -2,124 +2,138 @@
...
-fn exec_swap15(ref context: ExecutionContext) {
-    internal::exec_swap_i(ref context, 15);
-}
-
-/// 0x9F - SWAP16 operation
-/// Exchange 1st and 16th stack items.
-/// # Specification: https://www.evm.codes/#9f?fork=shanghai
-fn exec_swap16(ref context: ExecutionContext) {
-    internal::exec_swap_i(ref context, 16);
+    fn exec_swap_i(ref context: ExecutionContext, i: u8) -> Result<(), EVMError> {
```
Comment (danilowhk) [inline crates/evm/src/instructions/exchange_operations.cairo:19]: I think since there is already 

``context.stack.swap_i(i)`` this function becomes redundant.

File: crates/evm/src/instructions/exchange_operations.cairo [hunk 2]
```diff
@@ -2,124 +2,138 @@
...
+}
+
+#[generate_trait]
+impl ExchangeOperations of ExchangeOperationsTrait {
+    /// 0x90 - SWAP1 operation
+    /// Exchange 1st and 2nd stack items.
+    /// # Specification: https://www.evm.codes/#90?fork=shanghai
+
+    fn exec_swap1(ref self: ExecutionContext) {
+        internal::exec_swap_i(ref self, 1);
```
Comment (danilowhk) [inline crates/evm/src/instructions/exchange_operations.cairo:32]: use context.stack.swap_i(i) directly

File: crates/evm/src/instructions/exchange_operations.cairo [hunk 3]
```diff
@@ -2,124 +2,138 @@
...
+    }
+
+    /// 0x91 - SWAP2 operation
+    /// Exchange 1st and 3rd stack items.
+    /// # Specification: https://www.evm.codes/#91?fork=shanghai
+    fn exec_swap2(ref self: ExecutionContext) {
+        internal::exec_swap_i(ref self, 2);
```
Comment (danilowhk) [inline crates/evm/src/instructions/exchange_operations.cairo:39]: Same for all other swaps

File: crates/evm/src/instructions/exchange_operations.cairo [hunk 4]
```diff
@@ -2,124 +2,138 @@
...
+    }
+}
+
+#[generate_trait]
+impl ExchangeOperations of ExchangeOperationsTrait {
+    /// 0x90 - SWAP1 operation
+    /// Exchange 1st and 2nd stack items.
+    /// # Specification: https://www.evm.codes/#90?fork=shanghai
+
+    fn exec_swap1(ref self: ExecutionContext) {
```
Comment (Eikix) [inline crates/evm/src/instructions/exchange_operations.cairo:31]: Careful, your opcodes are not returning the required signature,
They need to return a `Result<(), EVMError>`.

File: crates/evm/src/instructions/exchange_operations.cairo (hunk 2 above)
Comment (Eikix) [reply crates/evm/src/instructions/exchange_operations.cairo:32]:
```suggestion
        context.stack.swap_i(i.into())
```

File: crates/evm/src/interpreter.cairo [hunk 5]
```diff
@@ -535,67 +535,67 @@ impl EVMInterpreterImpl of EVMInterpreterTrait {
         }
         if opcode == 144 {
             // SWAP1
-            exchange_operations::exec_swap1(ref context);
+            context.exec_swap1();
```
Comment (Eikix) [inline crates/evm/src/interpreter.cairo:538]:
```suggestion
            return context.exec_swap1();
```

File: crates/evm/src/interpreter.cairo [hunk 6]
```diff
@@ -535,67 +535,67 @@ impl EVMInterpreterImpl of EVMInterpreterTrait {
...
         }
         if opcode == 145 {
             // SWAP2
-            exchange_operations::exec_swap2(ref context);
+            context.exec_swap2();
```
Comment (Eikix) [inline crates/evm/src/interpreter.cairo:542]: same for all other swaps

File: crates/evm/src/tests/test_instructions/test_exchange_operations.cairo [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+
+
+#[test]
+#[available_gas(20000000)]
+fn test_swap1() {
+    let mut ctx = setup_execution_context();
+    ctx.stack.push(2).unwrap();
+    ctx.stack.push(3).unwrap();
+    ctx.exec_swap1();
+    assert(ctx.stack.peek().unwrap() == 2, 'Top should be now [1]');
```
Comment (Eikix) [inline crates/evm/src/tests/test_instructions/test_exchange_operations.cairo:17]:
```suggestion
    assert(ctx.stack.peek().unwrap() == 2, 'Top should be now 2');
```

File: crates/evm/src/tests/test_instructions/test_exchange_operations.cairo [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+}
+
+#[test]
+#[available_gas(20000000)]
+fn test_swap2() {
+    let mut ctx = setup_execution_context();
+    ctx.stack.push(2).unwrap();
```
Comment (Eikix) [inline crates/evm/src/tests/test_instructions/test_exchange_operations.cairo:24]: it's not so clear using these numbers, I'd advise using:
0x01 for the top of the stack element
0xff for the swapped element
0x00 for the elements in between

Comment (Eikix) [reply crates/evm/src/tests/test_instructions/test_exchange_operations.cairo:24]: 1
0
0
0
0
0
0xff

-> exec_swap6

->
0xff
0
0
0
0
0
1

**Review by Eikix — CHANGES_REQUESTED** [review]
Nice PR!

File: crates/evm/src/instructions/exchange_operations.cairo (hunk 1 above)
Comment (enitrat) [reply crates/evm/src/instructions/exchange_operations.cairo:19]: Yes, you can remove it

File: crates/evm/src/tests/test_instructions/test_exchange_operations.cairo [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+use evm::tests::test_utils::setup_execution_context;
+use array::ArrayTrait;
+use evm::context::ExecutionContextTrait;
+use evm::instructions::exchange_operations::ExchangeOperationsTrait;
+use core::result::ResultTrait;
+
+
+#[test]
+#[available_gas(20000000)]
+fn test_swap1() {
```
Comment (enitrat) [inline crates/evm/src/tests/test_instructions/test_exchange_operations.cairo:12]:
```suggestion
fn test_exec_swap1() {
```

Same for all other test names

**Review by enitrat — COMMENTED** [review]
Your swap operations need to return Result<(), EVMError>

They will actually just return whatever `self.stack.swap_i(<value>)` returns

File: crates/evm/src/tests/test_instructions/test_exchange_operations.cairo [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+
+#[test]
+#[available_gas(20000000)]
+fn test_swap2() {
+    let mut ctx = setup_execution_context();
+    ctx.stack.push(2).unwrap();
+    ctx.stack.push(3).unwrap();
+    ctx.stack.push(3).unwrap();
+    ctx.exec_swap2();
+    assert(ctx.stack.peek().unwrap() == 2, 'Top should be now[2]');
```
[the remark this answers is not in the document]
Comment (Noeljarillo) [reply crates/evm/src/tests/test_instructions/test_exchange_operations.cairo:28]: makes sense will do number!

File: crates/evm/src/tests/test_instructions/test_exchange_operations.cairo (hunk 8 above)
Comment (Noeljarillo) [reply crates/evm/src/tests/test_instructions/test_exchange_operations.cairo:24]: makes sense!

File: crates/evm/src/instructions/exchange_operations.cairo (hunk 1 above)
Comment (Noeljarillo) [reply crates/evm/src/instructions/exchange_operations.cairo:19]: https://github.com/Noeljarillo/kakarot-ssj/blob/feat/implement_0x90_-_SWAP1_Opcode/crates/evm/src/instructions/push_operations.cairo followed the implementation of push

**Comment by Noeljarillo** [comment]
Fixed Returns on function and op code call, aslo added suggested changes to tests.

I've tried delete swap_i() function but after lot of thinkering could not delete as calling context.stack.swap_i() from every swap_n gave me errors that could not find the function. But prototyped functions and PUSH implementation is by using swap_i()

File: crates/evm/src/instructions/exchange_operations.cairo (hunk 1 above)
Comment (enitrat) [reply crates/evm/src/instructions/exchange_operations.cairo:19]: For push it's different because we have an extra line to read data at `i`. Here it's not useful at all

**Review by enitrat — COMMENTED** [review]
> calling context.stack.swap_i() from every swap_n gave me errors that could not find the function

You have to call `self.stack.swap_i`, as the swap_i are methods of the ExecutionContext type, which is passed as a  `self` parameter

**Comment by Noeljarillo** [comment]
Understood, I'll implement the way you suggest!

**Comment by Noeljarillo** [comment]
Million thanks @enitrat I was referencing wrong the call to swap i by calling it from context.stack instead of self, now it works as intended!

File: crates/evm/src/tests/test_instructions/test_exchange_operations.cairo [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+#[test]
+#[available_gas(20000000)]
+fn test_exec_swap1() {
+    let mut ctx = setup_execution_context();
+    // given
+    ctx.stack.push(0xf).unwrap();
+    ctx.stack.push(1).unwrap();
+    ctx.exec_swap1();
+
+    assert(ctx.stack.peek().unwrap() == 0xf, 'Top should be now 0xf');
```
Comment (enitrat) [inline crates/evm/src/tests/test_instructions/test_exchange_operations.cairo:19]: you also need to check that value at index `i` (here, one) has been swapped correctly. you can use `peek_at` for that

**Review by enitrat — APPROVED** [review]
lgtm!

**Review by Eikix — APPROVED** [review]
lgtm 🥕

**Comment by enitrat** [comment]
Closes #302
Closes #303
Closes #304
Closes #305
Closes #306
Closes #307 
Closes #308 
Closes #309 
Closes #310 
Closes #311 
Closes #312
Closes #313 
Closes #313 
Closes #314 
Closes #315  
Closes #316 
Closes #317
